Track bytes stored per file type and include in org metrics

tw4l opened the pull request:
Fixes #1206

tw4l committed:
Add bytes stored per type to org and metrics

The org now tracks bytesStored by type of crawl, uploads, and browser profiles
in addition to the total, and returns these values in the org metrics endpoint.

A migration is added to precompute these values in existing deployments.

In addition, all /metrics storage values are now returned solely as bytes, as
the GB form wasn't being used in the frontend and is unnecessary.

tw4l committed:
Ignore duplicate code in migrations

tw4l committed:
Update tests and drop Bytes from type values in metrics

tw4l committed:
Inc bytesStoredCrawls after crawl files stored

tw4l committed:
Update OrgMetrics model for new field names

ikreymer commented on backend/btrixcloud/orgs.py:
```diff
@@ -330,17 +340,13 @@ async def get_max_concurrent_crawls(self, oid):
     async def add_crawl_files_to_org_bytes_stored(self, oid: uuid.UUID, size: int):
```
Wonder if this function redundant with the more general 'inc_org_bytes_stored' now?

tw4l replied:
Yup! I think it was a holdover from when we didn't have the ops methods available in the operator. Removed now.

ikreymer commented on backend/btrixcloud/orgs.py:
```diff
@@ -266,11 +264,23 @@ async def get_max_pages_per_crawl(self, oid: uuid.UUID):
         await self.orgs.find_one_and_update(
             {"_id": oid}, {"$inc": {"bytesStored": size}}
         )
+        if type_ == "crawl":
```
very small optimization: could combine it into a single query that's updated to get: `{"$inc": {"bytesStored": size, "bytesStoredCrawls": size}}`

ikreymer commented on backend/btrixcloud/basecrawls.py:
```diff
@@ -209,7 +209,9 @@ async def delete_crawls(
         res = await self.crawls.delete_many(query)
 
-        quota_reached = await self.orgs.inc_org_bytes_stored(org.id, -size)
+        if not type_:
```
Hm, i think maybe we can't do a bulk delete without type, since it could be a mixture of items?
Probably will need to do three deletes, so we know which quota to decrease?

tw4l replied:
I reworked `delete_crawls_all_types` to call the `delete_crawls` endpoint separately for crawls and uploads, and added in a missing routine to update the associated workflows and recompute the last crawl statistics. That routine also had a bug where it assumed only one crawl was being deleted per cid, so fixed that as well.

tw4l committed:
Fix test

tw4l committed:
Make code review revisions

- Remove redundant add_crawl_files_to_org_bytes_stored method
- Handle bulk deletes by type
- Fix bug where it was always assumed only one crawl was deleted per cid
and size was not tracked per cid
- Combine inc_org_bytes_stored into single query

tw4l committed:
Fixup

tw4l committed:
Make fixes to ensure workflow size is correct after deletion

tw4l committed:
Extend tests to test mixed type delete with stats

tw4l committed:
Add pylint disable comment

tw4l committed:
Fix test fixture

Chickensoupwithrice commented on backend/btrixcloud/migrations/migration_0017_storage_by_type.py:
```diff
@@ -0,0 +1,73 @@
+"""
+Migration 0015 - Calculate and store org storage usage by type
```
Should probably be `Migration 0017` but it's a comment so /shrug

tw4l replied:
Thanks, nice catch!

tw4l committed:
Fix migration number in docstring

Chickensoupwithrice reviewed (approved):
LGTM!

ikreymer commented on backend/btrixcloud/basecrawls.py:
```diff
@@ -496,13 +501,56 @@ async def delete_crawls_all_types(
+
+        deleted_count = 0
+
+        crawls_to_delete = await self._filter_delete_list_by_type(
```
This all works, but there's a more optimal way without having to do the filtering by type here actually.
Since the delete function already calls get_crawl_raw, we have the type available there, and it can just bin the crawls by type there.

tw4l replied:
See other comment - now splitting in one pass, but for now I think there are advantages to keeping `delete_crawls` to one type at a time.

ikreymer commented on backend/btrixcloud/basecrawls.py:
```diff
@@ -190,28 +190,33 @@ async def update_crawl(
 
         for crawl_id in delete_list.crawl_ids:
             crawl = await self.get_crawl_raw(crawl_id, org)
-            size += await self._delete_crawl_files(crawl, org)
```
Can filter by crawl type here, since we have the type available. Files are deleted for all crawls, but then can do:
```
deleted_size = await self._delete_crawl_files(crawl, org)
if crawl.type == "crawl":
   crawl_size += delete_size
else:
   upload_size += delete_size
```

ikreymer replied:
Actually, it looks like we're not even checking the type before deleting crawl files here!
Should probably do:
```
if type_ and crawl.type != type_:
   continue
```

This is actually a bug in the current version as well, where it'll delete the files regardless of type, but not actual archived item object!

tw4l replied:
Good catch on the missing type check! That's now been added.

I think there are advantages to handling the complexity of deleting multiple types in `delete_crawls_all_types` rather than `delete_crawls`. Everywhere else in our app that we are deleting content other than through the `/all-crawls` delete endpoint, we're handling one type at a time. And having `delete_crawls` handle only one type at a time keeps things a bit simpler, e.g. letting `inc_org_bytes_stored`only need to worry about one archived item type at a time.

I pushed a change to `delete_crawls_all_types` to split the delete list into crawls and uploads arrays in one pass rather than two, which should help.

There are other improvements to be made to the delete endpoints that I'm planning on doing in a separate pass in this sprint (see https://github.com/webrecorder/browsertrix-cloud/issues/1208), so I'd expect some of this to change and get optimized there, but I don't want to hold up this PR for something that's creeping a bit out of scope just because I discovered some bugs 😅 If that sounds reasonable to you!

ikreymer reviewed (changes requested):
On closer look, maybe we should fix at least the type checking bug before merging, so it doesn't bite us in the future..

tw4l committed:
Add type check within delete_crawls

tw4l committed:
Split crawls and uploads in one filter pass

tw4l committed:
Linting

tw4l commented:
> On closer look, maybe we should fix at least the type checking bug before merging, so it doesn't bite us in the future..

Type checking bug is fixed!

tw4l committed:
Add missing await and type hint

tw4l committed:
Fix getter for crawl type in type check
